### DISCUSSION TITLE
Opening a file manager in the current directory

### DISCUSSION THREAD

**Discussion by salmankhilji** [kovidgoyal/kitty#4608] [Q&A] [answered]
I would like to open a file manager of my choice within the cwd from within a kitty window using a key-binding like this:

```
map kitty_mod+p>p launch --type=background --cwd=current nautilus
```

The problem is that nautilus (on Ubuntu) does not start in the cwd.  Instead, it wants a command line argument.  I see that the `launch` command has special arguments like `@selection`, `@active-kitty-window-id` etc.  But I don't see something like `@cwd`.  Can it be added as a new feature?

I am also not sure if its worth it as we have the following workarounds:

1. I suppose I can write a wrapper script that launches nautilus taking a --cwd argument.  

2. bind the `nautilus .` within `.bashrc`.

**Comment by kovidgoyal** [comment, marked as answer]
If nautilus is silly enough to not respect the current working directory
then just use the following 

```
map kitty_mod+p>p launch --type=background --cwd=current sh -c "exec nautilus $PWD"
```

**Reply by salmankhilji** [reply]
Oddly enough, this does not work—`$PWD` always is set to `$HOME` for some reason.

To help debug this issue, I manually ran a slightly different command from the command line:

```sh
sh -c "notify-send $PWD"
```

The above correctly prints the cwd within the notification area on Ubuntu.  However the exact same command when put inside `kitty.conf` file and launched using the F9 key always prints the home directory.  So there must be something odd about the way kitty launches the program.

The following helper script, put somewhere in `$PATH` works:

```
#!/bin/sh
nautilus "$PWD"
```

However, I am reluctant to add a helper script to my system for something as silly as this.

**Comment by kovidgoyal** [comment]
Works fine for me with

kitty -o 'map f9 launch --cwd=current bash -c "notify-send $(pwd)"'

**Reply by salmankhilji** [reply]
> Works fine for me with kitty -o 'map f9 launch --cwd=current bash -c "notify-send $(pwd)"'

Yes, this works!  Its the `$(pwd)` part that made the difference.  `$PWD` does not work.

Here is the final line in my `kitty.conf`:

```sh
map f9 launch --type=background --cwd=current sh -c "nautilus $(pwd)"
```

**Comment by salmankhilji** [comment]
I am now wanting to change my approach a bit.

I realize that the following line my my `kitty.conf`, which works beautifully BTW, is more or less redundant with the default binding for `pass_selection_to_program`:

```
map f9 launch --type=background --cwd=current sh -c "xdg-open $(pwd)"
```

With the default mapping for `pass_selection_to_program`, I can open a _subfolder_ within my file manager by first double-clicking the _subfolder_ to select it and then `CTRL+SHIFT+O` to open it within the file manager.

Opening a file manager window within the _current_ directory, however, is not very convenient.  I have to first find a `.` (dot) character within the window, move my mouse to select it, and then `CTRL+SHIFT+O` to open it.

As a workaround, I tried:

#### kitty.conf

```
map kitty_mod+o pass_selection_to_program kitty-open-helper.sh
```

#### kitty-open-helper.sh
```sh
#!/bin/sh
[ $# -gt 0 ] || set -- .
exec xdg-open "$@"
```

But this does not work because if there is no selection, kitty does not invoke the program.

So I made the following changes to kitty itself to make it work:

```diff
diff --git a/kitty/window.py b/kitty/window.py
index 2ea964dc..da29639d 100644
--- a/kitty/window.py
+++ b/kitty/window.py
@@ -1471,11 +1471,11 @@ def copy_and_clear_or_interrupt(self) -> None:
     def pass_selection_to_program(self, *args: str) -> None:
         cwd = self.cwd_of_child
         text = self.text_for_selection()
-        if text:
-            if args:
-                open_cmd(args, text, cwd=cwd)
-            else:
-                open_url(text, cwd=cwd)
+        if not text: text = '.'
+        if args:
+            open_cmd(args, text, cwd=cwd)
+        else:
+            open_url(text, cwd=cwd)
 
     @ac('cp', 'Clear the current selection')
     def clear_selection(self) -> None:
```

Is there a better way to accomplish the same thing so that I can continue to use the `CTRL+SHIFT+O` shortcut to open a file manager window in pwd if there is no selection?

**Reply by page-down** [reply]
For now, in my opinion the best way to do this is to move the code into a custom kitten.
You have already seen that this will be very simple.

> ... if there is no selection, kitty does not invoke the program.

It looks like there could be a new option to invoke the program with or without a text selection and let the program decide for itself.

**Reply by salmankhilji** [reply]
This works:

#### kitty.conf

```
map kitty_mod+o launch --type=background --cwd=current kitty-open-helper.sh @selection
```

#### kitty-open-helper.sh

```sh
#!/bin/sh
[ "$1" = @selection ] && set -- .
xdg-open "$1"
```

I take advantage of the fact that if there is no selection, kitty calls the program with a sole argument `@selection`—I hope that this is by design and not a bug.

But I must be missing something—if it was _that_ simple, the `pass_selection_to_program` action wouldn't exist!

So what is it that `pass_selection_to_program` action provides that I can not accomplish with `launch --type=background --cwd=current`?

**Reply by page-down** [reply]
> ... I hope that this is by design and not a bug.
> ... if it was that simple ... the ... action wouldn't exist!
> ... what is it that ...  action provides that I can not accomplish with ...

I believe that all works as expected.
These actions are designed for easy configuration, for example `input_unicode_character` is very simple.
What they offer: no need to write many arguments. The disadvantage: not flexible enough.

So my previous suggestion doesn't hold up, because the added arguments increase the complexity.
